tests/src/main/java/org/hibernate/beanvalidation/tck/tests/AbstractBootstrapFailureTCKTest.java

```diff
@@ -0,0 +1,99 @@
+
+//CHECKSTYLE:OFF
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import org.testng.IHookCallBack;
+import org.testng.ITestResult;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.BeforeMethod;
+//CHECKSTYLE:ON
+
+/**
+ * This base class may only be used to test bootstrapping failures.
+ * <p>
+ * In some cases, the failure can either happen at deployment or at bootstrap. Thus, with this class, when the
+ * deployment is failing, we catch the deployment error if it matches a certain type of exception, and we consider it at
+ * the exception that will be thrown by the tests.
+ * <p>
+ * In this case, the content of the tests is not executed so you need to be very careful to use this class only when
+ * this behavior is required.
+ *
+ * @author Guillaume Smet
+ */
+public abstract class AbstractBootstrapFailureTCKTest extends AbstractTCKTest {
+
+	private Exception acceptedDeploymentException;
+
+	protected abstract Class<? extends Exception> acceptedDeploymentExceptionType();
+
+	@Override
+	@BeforeClass(groups = "arquillian", inheritGroups = true)
+	public void arquillianBeforeClass() throws Exception {
+		try {
+			super.arquillianBeforeClass();
+		}
+		catch (Exception e) {
+			// if the deployment throws an exception and if it's of the right type, we don't throw the exception here
+			// and we will "virtually" throw it instead of executing the tests
+			acceptedDeploymentException = extractAcceptedDeploymentException( acceptedDeploymentExceptionType(), e );
+			if ( acceptedDeploymentException == null ) {
+				throw e;
+			}
+		}
+	}
+
+	@Override
+	@BeforeMethod(groups = "arquillian", inheritGroups = true)
+	public void arquillianBeforeTest(Method testMethod) throws Exception {
+		// Don't do anything if the deployment failed
+		if ( acceptedDeploymentException != null ) {
+			return;
+		}
+
+		super.arquillianBeforeTest( testMethod );
+	}
+
+	@Override
+	public void run(IHookCallBack callback, ITestResult testResult) {
+		if ( acceptedDeploymentException != null ) {
+			// We consider that the test failed with the exception caught during the deployment
+			testResult.setStatus( ITestResult.FAILURE );
+			testResult.setThrowable( acceptedDeploymentException );
+
+			return;
+		}
+
+		super.run( callback, testResult );
+	}
+
+	private static Exception extractAcceptedDeploymentException(Class<? extends Exception> acceptedDeploymentExceptionType, Throwable exception) {
+		if ( acceptedDeploymentExceptionType.isAssignableFrom( exception.getClass() ) ) {
+			return (Exception) exception;
+		}
+
+		if ( exception instanceof InvocationTargetException ) {
+			Exception acceptedDeploymentException = extractAcceptedDeploymentException( acceptedDeploymentExceptionType,
+					( (InvocationTargetException) exception ).getTargetException() );
+			if ( acceptedDeploymentException != null ) {
+				return acceptedDeploymentException;
+			}
+		}
+
+		if ( exception.getCause() != null ) {
```

Comment:
Apologies, I should have included the complete stack trace in my OP.  Unfortunately this wont' work for the use case I am running into, because Weld sets any deployment exceptions (there may be >1) as the message of the DeploymentException, instead of as a cause.  

So I think in addition to looking for chained exceptions, this method should also check if the exception message contains `acceptedDeploymentExceptionType.getName()`

Here is the full stack trace, to illustrate what I'm talking about:
```
org.jboss.weld.exceptions.DefinitionException: Exception List with 1 exceptions:
Exception 0 :
javax.validation.ValidationException: HV000201: Unable to instantiate clock provider class org.hibernate.beanvalidation.tck.tests.xmlconfiguration.XmlDefinedClockProviderWithNoDefaultConstructor.
	at org.hibernate.validator.internal.xml.ValidationBootstrapParameters.setClockProvider(ValidationBootstrapParameters.java:241)
	at org.hibernate.validator.internal.xml.ValidationBootstrapParameters.<init>(ValidationBootstrapParameters.java:61)
	at org.hibernate.validator.internal.engine.ConfigurationImpl.parseValidationXml(ConfigurationImpl.java:512)
	at org.hibernate.validator.internal.engine.ConfigurationImpl.buildValidatorFactory(ConfigurationImpl.java:317)
	at org.hibernate.validator.cdi.ValidationExtension.<init>(ValidationExtension.java:119)
	at com.ibm.ws.beanvalidation.v20.cdi.internal.LibertyHibernateValidatorExtension.delegate(LibertyHibernateValidatorExtension.java:112)
	at com.ibm.ws.beanvalidation.v20.cdi.internal.LibertyHibernateValidatorExtension.beforeBeanDiscovery(LibertyHibernateValidatorExtension.java:132)
	at sun.reflect.GeneratedMethodAccessor124.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.weld.injection.StaticMethodInjectionPoint.invoke(StaticMethodInjectionPoint.java:95)
	at org.jboss.weld.injection.MethodInvocationStrategy$SpecialParamPlusBeanManagerStrategy.invoke(MethodInvocationStrategy.java:144)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:330)
	at org.jboss.weld.event.ExtensionObserverMethodImpl.sendEvent(ExtensionObserverMethodImpl.java:123)
	at org.jboss.weld.event.ObserverMethodImpl.sendEvent(ObserverMethodImpl.java:308)
	at org.jboss.weld.event.ObserverMethodImpl.notify(ObserverMethodImpl.java:286)
	at javax.enterprise.inject.spi.ObserverMethod.notify(ObserverMethod.java:124)
	at org.jboss.weld.util.Observers.notify(Observers.java:166)
	at org.jboss.weld.event.ObserverNotifier.notifySyncObservers(ObserverNotifier.java:285)
	at org.jboss.weld.event.ObserverNotifier.notify(ObserverNotifier.java:273)
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:177)
	at org.jboss.weld.event.ObserverNotifier.fireEvent(ObserverNotifier.java:171)
	at org.jboss.weld.bootstrap.events.AbstractContainerEvent.fire(AbstractContainerEvent.java:53)
	at org.jboss.weld.bootstrap.events.AbstractDefinitionContainerEvent.fire(AbstractDefinitionContainerEvent.java:44)
	at org.jboss.weld.bootstrap.events.BeforeBeanDiscoveryImpl.fire(BeforeBeanDiscoveryImpl.java:54)
	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:391)
	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:76)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:142)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:139)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.ibm.ws.cdi.impl.CDIContainerImpl.applicationStarting(CDIContainerImpl.java:139)
	at com.ibm.ws.cdi.liberty.CDIRuntimeImpl.applicationStarting(CDIRuntimeImpl.java:383)
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:28)
	at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStarting(StateChangeServiceImpl.java:50)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.preDeployApp(DeployedAppInfoBase.java:383)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.deployApp(DeployedAppInfoBase.java:412)
	at com.ibm.ws.app.manager.war.internal.WARApplicationHandlerImpl.install(WARApplicationHandlerImpl.java:65)
	at com.ibm.ws.app.manager.internal.statemachine.StartAction.execute(StartAction.java:140)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1258)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.performAction(ApplicationStateMachineImpl.java:1106)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.run(ApplicationStateMachineImpl.java:881)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: javax.validation.ValidationException: HV000065: Unable to load class: org.hibernate.beanvalidation.tck.tests.xmlconfiguration.XmlDefinedClockProviderWithNoDefaultConstructor from org.eclipse.osgi.internal.loader.EquinoxClassLoader@169014c7[com.ibm.ws.org.hibernate.validator.6.0.4.Final:1.0.19.master(id=117)].
	at org.hibernate.validator.internal.util.privilegedactions.LoadClass.loadNonValidatorClass(LoadClass.java:132)
	at org.hibernate.validator.internal.util.privilegedactions.LoadClass.run(LoadClass.java:64)
	at org.hibernate.validator.internal.util.privilegedactions.LoadClass.run(LoadClass.java:29)
	at org.hibernate.validator.internal.xml.ValidationBootstrapParameters.run(ValidationBootstrapParameters.java:298)
	at org.hibernate.validator.internal.xml.ValidationBootstrapParameters.setClockProvider(ValidationBootstrapParameters.java:234)
	... 43 more
Caused by: java.lang.ClassNotFoundException: org.hibernate.beanvalidation.tck.tests.xmlconfiguration.XmlDefinedClockProviderWithNoDefaultConstructor cannot be found by com.ibm.ws.org.hibernate.validator.6.0.4.Final_1.0.19.master
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:484)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:395)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:387)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:150)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at org.hibernate.validator.internal.util.privilegedactions.LoadClass.loadNonValidatorClass(LoadClass.java:129)
	... 47 more

	at org.jboss.weld.bootstrap.events.AbstractDefinitionContainerEvent.fire(AbstractDefinitionContainerEvent.java:46)
	at org.jboss.weld.bootstrap.events.BeforeBeanDiscoveryImpl.fire(BeforeBeanDiscoveryImpl.java:54)
	at org.jboss.weld.bootstrap.WeldStartup.startInitialization(WeldStartup.java:391)
	at org.jboss.weld.bootstrap.WeldBootstrap.startInitialization(WeldBootstrap.java:76)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:142)
	at com.ibm.ws.cdi.impl.CDIContainerImpl$2.run(CDIContainerImpl.java:139)
	at java.security.AccessController.doPrivileged(Native Method)
	at com.ibm.ws.cdi.impl.CDIContainerImpl.applicationStarting(CDIContainerImpl.java:139)
	at com.ibm.ws.cdi.liberty.CDIRuntimeImpl.applicationStarting(CDIRuntimeImpl.java:383)
	at com.ibm.ws.container.service.state.internal.ApplicationStateManager.fireStarting(ApplicationStateManager.java:28)
	at com.ibm.ws.container.service.state.internal.StateChangeServiceImpl.fireApplicationStarting(StateChangeServiceImpl.java:50)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.preDeployApp(DeployedAppInfoBase.java:383)
	at com.ibm.ws.app.manager.module.internal.DeployedAppInfoBase.deployApp(DeployedAppInfoBase.java:412)
	at com.ibm.ws.app.manager.war.internal.WARApplicationHandlerImpl.install(WARApplicationHandlerImpl.java:65)
	at com.ibm.ws.app.manager.internal.statemachine.StartAction.execute(StartAction.java:140)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.enterState(ApplicationStateMachineImpl.java:1258)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.performAction(ApplicationStateMachineImpl.java:1106)
	at com.ibm.ws.app.manager.internal.statemachine.ApplicationStateMachineImpl.run(ApplicationStateMachineImpl.java:881)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Reply:
Hum. Can't say I'm excited about it :).

The issue is that I will have to build a fake exception so we won't be able to test the real content of the exception. But, thinking about it, I don't think we test it. I'll push something shortly.

Reply:
Agreed that it's a bit odd, but in a real world situation deployment exceptions aren't actually catchable by java code, so it doesn't really matter if this data is in the exception message or as a proper chained exception from the user perspective.

Thanks for looking into this so promptly btw!  I'm working on collecting the full set of classes effected now.

Reply:
BTW, in your case, you only have one exception in the list? Wondering if we should open an issue against Weld to better deal with this case.

Reply:
@aguibert pushed an additional commit to work around the Weld limitation

Didn't have the time to test it properly though (I have an appointment and need to go). Feel free to test it if you have time and it helps. Otherwise, I'll test it tomorrow.

If it doesn't work, feel free to adjust the approach. I think it's the right one, even if not perfect.